synapse/http/servlet.py

```diff
@@ -413,6 +415,69 @@ def parse_string(
     )
 
 
+EnumT = TypeVar("EnumT", bound=enum.Enum)
+
+
+@overload
+def parse_enum(
+    request: Request,
+    name: str,
+    E: Type[EnumT],
+    default: EnumT,
+) -> EnumT:
+    ...
+
+
+@overload
+def parse_enum(
+    request: Request,
+    name: str,
+    E: Type[EnumT],
+    *,
+    required: Literal[True],
+) -> EnumT:
+    ...
+
+
+def parse_enum(
+    request: Request,
+    name: str,
+    E: Type[EnumT],
+    default: Optional[EnumT] = None,
+    required: bool = False,
+) -> Optional[EnumT]:
+    """
+    Parse an enum parameter from the request query string.
+
+    Args:
+        request: the twisted HTTP request.
+        name: the name of the query parameter.
+        E: the enum which represents valid values
+        default: enum value to use if the parameter is absent, defaults to None.
+        required: whether to raise a 400 SynapseError if the
+            parameter is absent, defaults to False.
+
+    Returns:
+        An enum value.
+
+    Raises:
+        SynapseError if the parameter is absent and required, or if the
+            parameter is present, must be one of a list of allowed values and
+            is not one of those allowed values.
+    """
+    # TODO Assert the enum values are strings.
```

Comment:
We can likely get away without supporting numeric enums. I don't think they ever show up in query parameters? (not suggesting any change)

Reply:
I don't think we ever use numeric *enums* in query parameters, so I think that's fine.

Do you think it is worth trying to assert here? Note that using a `StrEnum` didn't work. (mypy didn't know what it was.)

Reply:
I don't mind either way. Not sure how we'd do the assert though, since enums can have a mix of string and number values in theory.

Reply:
Seems a little inefficient (not that it matters). I'd be in favour of a comment instead of the assert.
As long as we have test coverage for all our routes, CI ought to fail if we try to use a numeric enum here anyway.

Reply:
I think I'm going to add it, we should only be iterating over a few values and it seems better to error immediately.